File: language/compiler/src/com/kubadziworski/parsing/visitor/statement/StatementVisitor.java
package com.kubadziworski.parsing.visitor.statement;

import jiva.parser.JivaBaseVisitor;
import jiva.parser.JivaParser;
import jiva.parser.JivaParser.AddContext;
import jiva.parser.JivaParser.BlockContext;
import jiva.parser.JivaParser.ConditionalExpressionContext;
import jiva.parser.JivaParser.ConstructorCallContext;
import jiva.parser.JivaParser.DivideContext;
import jiva.parser.JivaParser.FunctionCallContext;
import jiva.parser.JivaParser.IfStatementContext;
import jiva.parser.JivaParser.MultiplyContext;
import jiva.parser.JivaParser.PrintStatementContext;
import jiva.parser.JivaParser.ReturnVoidContext;
import jiva.parser.JivaParser.ReturnWithValueContext;
import jiva.parser.JivaParser.SubstractContext;
import jiva.parser.JivaParser.SupercallContext;
import jiva.parser.JivaParser.ValueContext;
import jiva.parser.JivaParser.VarReferenceContext;
import jiva.parser.JivaParser.VariableDeclarationContext;
import com.kubadziworski.domain.node.expression.ConditionalExpression;
import com.kubadziworski.domain.node.expression.Expression;
import com.kubadziworski.domain.scope.Scope;
import com.kubadziworski.domain.node.statement.*;
import com.kubadziworski.parsing.visitor.expression.ExpressionVisitor;
import org.antlr.v4.runtime.misc.NotNull;

/**
 * Created by kuba on 01.04.16.
 */

public class StatementVisitor extends JivaBaseVisitor<Statement> {

    private final ExpressionVisitor expressionVisitor;
    private final PrintStatementVisitor printStatementVisitor;
    private final VariableDeclarationStatementVisitor variableDeclarationStatementVisitor;
    private final ReturnStatementVisitor returnStatementVisitor;
    private final BlockStatementVisitor blockStatementVisitor;
    private final IfStatementVisitor ifStatementVisitor;
    private final ForStatementVisitor forStatementVisitor;
    private final AssignmentStatementVisitor assignmentStatementVisitor;

    public StatementVisitor(Scope scope) {
        expressionVisitor = new ExpressionVisitor(scope);
        printStatementVisitor = new PrintStatementVisitor(expressionVisitor);
        variableDeclarationStatementVisitor = new VariableDeclarationStatementVisitor(expressionVisitor,scope);
        returnStatementVisitor = new ReturnStatementVisitor(expressionVisitor);
        blockStatementVisitor = new BlockStatementVisitor(scope);
        ifStatementVisitor = new IfStatementVisitor(this,expressionVisitor);
        forStatementVisitor = new ForStatementVisitor(scope);
        assignmentStatementVisitor = new AssignmentStatementVisitor(expressionVisitor);
    }

    @Override
    public Statement visitPrintStatement(@NotNull PrintStatementContext ctx) {
        return printStatementVisitor.visitPrintStatement(ctx);
    }

    @Override
    public Statement visitVariableDeclaration(@NotNull VariableDeclarationContext ctx) {
        return variableDeclarationStatementVisitor.visitVariableDeclaration(ctx);
    }

    @Override
    public Statement visitReturnVoid(@NotNull ReturnVoidContext ctx) {
        return returnStatementVisitor.visitReturnVoid(ctx);
    }

    @Override
    public Statement visitReturnWithValue(@NotNull ReturnWithValueContext ctx) {
        return returnStatementVisitor.visitReturnWithValue(ctx);
    }

    @Override
    public Statement visitBlock(@NotNull BlockContext ctx) {
        return blockStatementVisitor.visitBlock(ctx);
    }

    @Override
    public Statement visitIfStatement(@NotNull IfStatementContext ctx) {
        return ifStatementVisitor.visitIfStatement(ctx);
    }

    @Override
    public Expression visitVarReference(@NotNull VarReferenceContext ctx) {
        return expressionVisitor.visitVarReference(ctx);
    }

    @Override
    public Expression visitValue(@NotNull ValueContext ctx) {
        return expressionVisitor.visitValue(ctx);
    }

    @Override
    public Expression visitFunctionCall(@NotNull FunctionCallContext ctx) {
        return expressionVisitor.visitFunctionCall(ctx);
    }

    @Override
    public Expression visitConstructorCall(@NotNull ConstructorCallContext ctx) {
        return expressionVisitor.visitConstructorCall(ctx);
    }

    @Override
    public Expression visitSupercall(@NotNull SupercallContext ctx) {
        return expressionVisitor.visitSupercall(ctx);
    }

    @Override
    public Expression visitAdd(@NotNull AddContext ctx) {
        return expressionVisitor.visitAdd(ctx);
    }

    @Override
    public Expression visitMultiply(@NotNull MultiplyContext ctx) {
        return expressionVisitor.visitMultiply(ctx);
    }

    @Override
    public Expression visitSubstract(@NotNull SubstractContext ctx) {
        return expressionVisitor.visitSubstract(ctx);
    }

    @Override
    public Expression visitDivide(@NotNull DivideContext ctx) {
        return expressionVisitor.visitDivide(ctx);
    }

    @Override
    public ConditionalExpression visitConditionalExpression(@NotNull ConditionalExpressionContext ctx) {
        return expressionVisitor.visitConditionalExpression(ctx);
    }

    @Override
    public Statement visitForStatement(@NotNull JivaParser.ForStatementContext ctx) {
        return forStatementVisitor.visitForStatement(ctx);
    }

    @Override
    public Statement visitAssignment(@NotNull JivaParser.AssignmentContext ctx) {
        return assignmentStatementVisitor.visitAssignment(ctx);
    }
}

File: language/compiler/src/parsing/visitor/statement/StatementVisitor.kt
package jiva.parsing.visitor.statement

import jiva.domain.node.expression.ConditionalExpression
import jiva.domain.node.expression.Expression
import jiva.domain.node.statement.Statement
import jiva.domain.scope.Scope
import jiva.parsing.visitor.expression.ExpressionVisitor
import jiva.parser.JivaBaseVisitor
import jiva.parser.JivaParser.*
import org.antlr.v4.runtime.misc.NotNull

/**
 * Created by kuba on 01.04.16.
 */
class StatementVisitor(scope: Scope) : JivaBaseVisitor<Statement?>() {

    private val expressionVisitor = ExpressionVisitor(scope)
    private val printStatementVisitor = PrintStatementVisitor(expressionVisitor)
    private val variableDeclarationStatementVisitor = VariableDeclarationStatementVisitor(expressionVisitor, scope)
    private val returnStatementVisitor = ReturnStatementVisitor(expressionVisitor)
    private val blockStatementVisitor = BlockStatementVisitor(scope)
    private val ifStatementVisitor = IfStatementVisitor(this, expressionVisitor)
    private val forStatementVisitor = ForStatementVisitor(scope)
    private val assignmentStatementVisitor = AssignmentStatementVisitor(expressionVisitor)

    override fun visitPrintStatement(ctx: PrintStatementContext): Statement {
        return printStatementVisitor.visitPrintStatement(ctx)
    }

    override fun visitVariableDeclaration(ctx: VariableDeclarationContext): Statement {
        return variableDeclarationStatementVisitor.visitVariableDeclaration(ctx)
    }

    override fun visitReturnVoid(ctx: ReturnVoidContext): Statement {
        return returnStatementVisitor.visitReturnVoid(ctx)
    }

    override fun visitReturnWithValue(ctx: ReturnWithValueContext): Statement {
        return returnStatementVisitor.visitReturnWithValue(ctx)
    }

    override fun visitBlock(ctx: BlockContext): Statement {
        return blockStatementVisitor.visitBlock(ctx)
    }

    override fun visitIfStatement(ctx: IfStatementContext): Statement {
        return ifStatementVisitor.visitIfStatement(ctx)
    }

    override fun visitVarReference(ctx: VarReferenceContext): Expression {
        return expressionVisitor.visitVarReference(ctx)
    }

    override fun visitValue(ctx: ValueContext): Expression {
        return expressionVisitor.visitValue(ctx)
    }

    override fun visitFunctionCall(ctx: FunctionCallContext): Expression {
        return expressionVisitor.visitFunctionCall(ctx)
    }

    override fun visitConstructorCall(ctx: ConstructorCallContext): Expression {
        return expressionVisitor.visitConstructorCall(ctx)
    }

    override fun visitSupercall(ctx: SupercallContext): Expression {
        return expressionVisitor.visitSupercall(ctx)
    }

    override fun visitAdd(ctx: AddContext): Expression {
        return expressionVisitor.visitAdd(ctx)
    }

    override fun visitMultiply(ctx: MultiplyContext): Expression {
        return expressionVisitor.visitMultiply(ctx)
    }

    override fun visitSubstract(ctx: SubstractContext): Expression {
        return expressionVisitor.visitSubstract(ctx)
    }

    override fun visitDivide(ctx: DivideContext): Expression {
        return expressionVisitor.visitDivide(ctx)
    }

    override fun visitConditionalExpression(ctx: ConditionalExpressionContext): ConditionalExpression {
        return expressionVisitor.visitConditionalExpression(ctx)
    }

    override fun visitForStatement(ctx: ForStatementContext): Statement {
        return forStatementVisitor.visitForStatement(ctx)
    }

    override fun visitAssignment(ctx: AssignmentContext): Statement {
        return assignmentStatementVisitor.visitAssignment(ctx)
    }

}